tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java

```diff
@@ -111,10 +117,28 @@ private void commitTransaction(GlobalTransaction tx) throws TransactionalExecuto
 
     private void rollbackTransaction(GlobalTransaction tx, Throwable ex) throws TransactionException, TransactionalExecutor.ExecutionException {
         triggerBeforeRollback();
-        tx.rollback();
+        //branch has reported rollback done
+        if (!(ex instanceof TransactionException
+                && ((TransactionException) ex).getCode() == TransactionExceptionCode.ParticipantRollbackDone)) {
+            tx.rollback();
+        }
         triggerAfterRollback();
         // 3.1 Successfully rolled back
-        throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex);
+        if (tx.getRole() == GlobalTransactionRole.Launcher) {
+            if (ex instanceof TransactionException) {
+                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex.getCause());
+            } else {
+                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex);
+            }
+        } else if (tx.getRole() == GlobalTransactionRole.Participant) {
+            if (ex instanceof TransactionException) {
+                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone,
+                        new TransactionException(TransactionExceptionCode.ParticipantRollbackDone, ex.getCause()));
```

Comment:
ex.getCause()==ex
I think that only the code is different here, you can write a get code function, and completeTransactionAfterThrowing is also.

Reply:
> ex.getCause()==ex
When using the construction method，ex.geCause() == originalException